fluentd/run.sh

```diff
@@ -107,32 +105,89 @@ else
-if [ $BUFFER_SIZE_LIMIT -eq 0 ]; then
-    BUFFER_SIZE_LIMIT=1048576
+
+if [ "$BUFFER_TYPE" = "file" -a -z "$FILE_BUFFER_PATH" ]; then
+    echo "ERROR: BUFFER_TYPE is set to file but FILE_BUFFER_PATH is empty.  Please specify a FILE_BUFFER_PATH.  Exiting."
 fi
 
-DIV=1
-if [ "$ES_HOST" != "$OPS_HOST" ] || [ "$ES_PORT" != "$OPS_PORT" ] ; then
-    # using ops cluster
-    DIV=`expr $DIV \* 2`
+FILE_BUFFER_PATH_ES=""
+FILE_BUFFER_PATH_ESOPS=""
+FILE_BUFFER_PATH_ES_COPY=""
+FILE_BUFFER_PATH_ESOPS_COPY=""
+# Default buffer_type: file if FILE_BUFFER_PATH is specified.
+if [ "${BUFFER_TYPE:-}" = "file" ]; then
+    if [ ! -d $FILE_BUFFER_PATH ]; then
+        mkdir -p $FILE_BUFFER_PATH
+    fi
+    # Get the available disk size; use 1/4 of it
+    DF_LIMIT=$(df -B1 $FILE_BUFFER_PATH | grep -v Filesystem | awk '{print $2}')
+    DF_LIMIT=${DF_LIMIT:-0}
+    DF_LIMIT=$(expr $DF_LIMIT / 4) || :
+    if [ $DF_LIMIT -eq 0 ]; then
+        echo "ERROR: No disk space is available for file buffer in $FILE_BUFFER_PATH.  Using memory buffer."
+        exit 1
+    fi
+    # Given buffer limit per output
+    TOTAL_LIMIT=$(echo ${FILE_BUFFER_LIMIT:-0} | sed -e "s/[Kk]/*1024/g;s/[Mm]/*1024*1024/g;s/[Gg]/*1024*1024*1024/g;s/i//g" | bc) || :
+    # In total
+    TOTAL_LIMIT=$(expr $TOTAL_LIMIT \* $DIV) || :
+    if [ $TOTAL_LIMIT -le 0 ]; then
+        echo "ERROR: Invalid file buffer limit ($FILE_BUFFER_LIMIT) is given.  Failed to convert to bytes."
+        exit 1
+    fi
+    # Available disk space is less than the specified size.
+    if [ $DF_LIMIT -lt $TOTAL_LIMIT ]; then
+        echo "WARNING: Available disk space ($DF_LIMIT bytes) is less than the user specified file buffer limit ($FILE_BUFFER_LIMIT)."
+        TOTAL_LIMIT=$DF_LIMIT
+    fi
+    FILE_BUFFER_PATH_ES=${FILE_BUFFER_PATH}/es
+    FILE_BUFFER_PATH_ESOPS=${FILE_BUFFER_PATH}/esops
+    FILE_BUFFER_PATH_ES_COPY=${FILE_BUFFER_PATH}/escopy
+    FILE_BUFFER_PATH_ESOPS_COPY=${FILE_BUFFER_PATH}/esopscopy
+else
+    BUFFER_TYPE="memory"
 fi
-if [ "${USE_MUX_CLIENT:-}" = "true" ] ; then
-    DIV=`expr $DIV \* 2`
+if [ "$BUFFER_TYPE" = "memory" ] ; then
+    if [ "${USE_MUX:-}" = "true" ] ; then
+        TOTAL_LIMIT=$(echo $MUX_MEMORY_LIMIT | sed -e "s/[Kk]/*1024/g;s/[Mm]/*1024*1024/g;s/[Gg]/*1024*1024*1024/g;s/i//g" | bc) || :
+    else
+        TOTAL_LIMIT=$(echo $FLUENTD_MEMORY_LIMIT | sed -e "s/[Kk]/*1024/g;s/[Mm]/*1024*1024/g;s/[Gg]/*1024*1024*1024/g;s/i//g" | bc) || :
+    fi
+    # Use 75% of the avialble memory
+    TOTAL_LIMIT=$(expr $(expr $TOTAL_LIMIT \* 3) / 4) || :
 fi
+if [ -z $TOTAL_LIMIT -o $TOTAL_LIMIT -eq 0 ]; then
+    if [ "$USE_MUX" = "true" ]; then
+        # 75 % of 2G
+        TOTAL_LIMIT=1610612736
+    else
+        # 75 % of 512M
+        TOTAL_LIMIT=402653184
+    fi
+fi
+
+BUFFER_SIZE_LIMIT=$(echo $BUFFER_SIZE_LIMIT |  sed -e "s/[Kk]/*1024/g;s/[Mm]/*1024*1024/g;s/[Gg]/*1024*1024*1024/g;s/i//g" | bc)
+BUFFER_SIZE_LIMIT=${BUFFER_SIZE_LIMIT:-1048576}
```

Comment:
This calculation now needs to account for how many tags we have in flight when tags land in the elasticsearch output plugins.  I think we need to consider a change to tags first so that this calculation will be correct.